src/core/net/ip6.cpp

```diff
@@ -1007,45 +1005,62 @@ const NetifUnicastAddress *Ip6::SelectSourceAddress(MessageInfo &aMessageInfo)
                 // Rule 2: Prefer appropriate scope
                 if (addr->GetScope() >= overrideScope)
                 {
-                    rvalAddr = addr;
-                    rvalIface = candidateId;
-                    rvalPrefixMatched = candidatePrefixMatched;
+                    goto replace;
```

Comment:
@jwhui Using inline would involve a lot of context variables, `interfaceid`, `destinationScope`, etc. So I think we can just repeat those codes to avoid goto. Thanks for your suggestions.